### PR TITLE
CounterListType: add List for easy extending counters

### PR DESCRIPTION
**Pull request by Hanmac**
Part of #10076 

adds an easy extendable List for Counters where colors can be defined

### CODE REVIEW

**Comment by Hanmac** [comment]
Might be interesting if we can add some Icons for such List Counters too

like MTGO has Icons for these: https://mtg.wiki/page/Page_counter

**Comment by Hanmac** [comment]
@tool4ever most of the remaining enum types are used for AI in `isNegativeCounter` and `isUselessCounter`

i might try to move them into another text files too, or what's your opinion?

**Comment by tool4ever** [comment]
hmn, binding them to Enum to enforce the correct types at compile time has some advantages like not having to worry about String typos or them drifting apart 🤔

I'm not sure giving that up would be worth it?

File: forge-game/src/main/java/forge/game/card/CounterType.java [hunk 1]
```diff
@@ -16,12 +16,17 @@ static CounterType getType(String name) {
         try {
             return CounterEnumType.getType(name);
         } catch (final IllegalArgumentException ex) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/CounterType.java:18]: I hate this control flow by exception handling anti-pattern:
we should really try to get rid of it instead of relying even more on it

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/CounterType.java:18]: That's just the way for Enums

even apache-commons use it for their EnumUtils:
https://github.com/apache/commons-lang/blob/71bd23bb9ec4d27d314ca80ad289821e42e41f72/src/main/java/org/apache/commons/lang3/EnumUtils.java#L219-L228

**Comment by Hanmac** [comment]
> hmn, binding them to Enum to enforce the correct types at compile time has some advantages like not having to worry about String typos or them drifting apart 🤔
> 
> I'm not sure giving that up would be worth it?

Typos already happens anyway, they are just treated as CounterKeywordType which is the fallback anyway

also, we can't really support custom-types without allowing typos

File: forge-game/src/main/java/forge/game/card/CounterEnumType.java [hunk 2]
```diff
@@ -33,240 +30,42 @@ public enum CounterEnumType implements CounterType {
...
-
-    FUNK("FUNK", 215, 24, 222),
-
-    FURY("FURY", 255, 120, 89),
-
-    FUSE("FUSE", 255, 122, 85),
-
-    GEM("GEM", 255, 99, 251),
-
     GHOSTFORM("GHSTF", 223, 0, 254),
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/CounterEnumType.java]: the current split is still too weird imo

to make more sense only rules based ones should remain here, otherwise you can never be sure if it's already defined in either place...

and then some comment on this class should explain that

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/CounterEnumType.java]: @tool4ever yeah, i need to find a way to extract Negative and Useless counters into extra lists.
Or maybe add `[useless]` grouping into the new List file.

What's your opinion?

Comment (tool4ever) [reply forge-game/src/main/java/forge/game/card/CounterEnumType.java]: hmn you mean AI logic in a text file? don't think that really works, as some need checks for boardstate context anyway

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/CounterEnumType.java]: @tool4ever i mean these:
https://github.com/Card-Forge/forge/blob/ec6f1a6a98c2d3e4a4c26e9e3c92d01f3c231eab/forge-ai/src/main/java/forge/ai/ComputerUtil.java#L2874-L2875

and these for `isNegativeCounter`:
https://github.com/Card-Forge/forge/blob/ec6f1a6a98c2d3e4a4c26e9e3c92d01f3c231eab/forge-ai/src/main/java/forge/ai/ComputerUtil.java#L2852-L2856

The ones where we (unconditionally) know that they are useless/negative.
And if possible, i want these lists to be extendable too.

Some counters still might have some logic:
https://github.com/Card-Forge/forge/blob/ec6f1a6a98c2d3e4a4c26e9e3c92d01f3c231eab/forge-ai/src/main/java/forge/ai/ComputerUtil.java#L2844-L2851

**Comment by Hanmac** [comment]
List to be implemented:

```
[Positive]
ACORN=ACORN,139,69,19
ACQUIREDTASTE=acquired taste,139,19,69
AEGIS=AEGIS,207,207,207
AIM=AIM,255,180,0
ARROW=ARROW,237,195,0
ARROWHEAD=ARWHD,230,191,167
BAIT=BAIT,120,100,60
BLAZE=BLAZE,255,124,82
BLESSING=BLESS,251,0,94
BLIGHT=BLGHT,130,115,160
BLOOD=BLOOD,255,108,111
BLOODLINE=BLDLN,224,44,44
BLOODSTAIN=BLDST,224,44,44
BORE=BORE,98,47,34
BOUNTY=BOUNT,255,158,0
BRAIN=BRAIN,197,62,212
BRICK=BRICK,226,192,164
BURDEN=BURDEN,135,62,35
CAGE=CAGE,155,155,155
CARRION=CRRON,255,163,222
CELL=CELL,90,10,95
CHARGE=CHARG,246,192,0
CHORUS=CHRUS,0,192,246
COIN=COIN,255,215,0
COLLECTION=CLLCT,255,215,0
COMPONENT=COMPN,224,160,48
CONQUEROR=CONQR,225,210,25
CONTESTED=CONTES,255,76,2
CORPSE=CRPSE,230,186,209
CORRUPTION=CRPTN,210,121,210
CREDIT=CRDIT,188,197,234
CROAK=CROAK,155,255,5
CRYSTAL=CRYST,255,85,206
CUBE=CUBE,148,219,0
CURRENCY=CURR,223,200,0
DEATH=DEATH,255,108,110
DEFENSE=DEF,164,23,32
DELAY=DELAY,102,206,255
DEPLETION=DPLT,185,201,208
DESCENT=DESCT,175,35,40
DESPAIR=DESPR,238,186,187
DEVOTION=DEVOT,255,111,255
DISCOVERY=DISCO,12,230,100
DIVINITY=DVNTY,0,233,255
DREAD=DREAD,205,170,240
DREAM=DREAM,190,189,255
DUTY=DUTY,232,245,245
ECHO=ECHO,225,180,255
EGG=EGG,255,245,195
ELIXIR=ELIXR,81,221,175
EMBER=EMBER,247,52,43
EON=EON,23,194,255
ERUPTION=ERUPTION,255,124,124
EVERYTHING=EVRY,255,255,255
EXPOSURE=EXPOSURE,50,180,30
EYEBALL=EYE,184,202,201
EYESTALK=EYES,184,202,201
FADE=FADE,159,209,192
FATE=FATE,255,164,226
FEATHER=FTHR,195,202,165
FEEDING=FEED,245,21,5
FELLOWSHIP=FLWS,255,255,255
FETCH=FETCH,180,235,52
FILIBUSTER=FLBTR,255,179,119
FILM=FILM,255,255,255
FIRE=FIRE,240,30,35
FLAME=FLAME,255,143,43
FLAVOR=FLAVOR,208,152,97
FLOOD=FLOOD,0,203,255
FORESHADOW=FRSHD,144,99,207
FUNGUS=FNGUS,121,219,151
FUNK=FUNK,215,24,222
FURY=FURY,255,120,89
FUSE=FUSE,255,122,85
GEM=GEM,255,99,251
GLYPH=GLYPH,184,202,199
GROWTH=GRWTH,87,226,32
HARMONY=HRMNY,0,230,155
HATCHING=HATCH,204,255,204
HATCHLING=HTCHL,201,199,186
HEALING=HEAL,255,166,236
HIT=HIT,255,245,195
HONE=HONE,51,227,255
HOOFPRINT=HOOF,233,189,170
HOPE=HOPE,232,245,245
HOURGLASS=HRGLS,0,215,255
HOUR=HOUR,198,197,210
HUNGER=HUNGR,255,91,149
HUSK=HUSK,227,212,173
ICE=ICE,0,239,255
IMPOSTOR=IMPO,173,194,255
INCUBATION=INCBT,40,210,25
INFECTION=INFCT,0,230,66
INFLUENCE=INFL,201,99,212
INGENUITY=INGTY,67,186,205
INGREDIENT=INGRD,180,50,145
INTEL=INTEL,80,250,180
INTERVENTION=INTRV,205,203,105
INVITATION=INVIT,205,0,26
ISOLATION=ISOLT,250,190,0
JAVELIN=JAVLN,180,206,172
JUDGMENT=JUDGM,249,220,52
KICK=KICK,255,255,240
KI=KI,190,189,255
KNOWLEDGE=KNOWL,0,115,255
LANDMARK=LNMRK,186,28,28
LEVEL=LEVEL,60,222,185
LOOT=LOOT,255,215,0
LORE=LORE,209,198,161
LOYALTY=LOYAL,198,198,198
LUCK=LUCK,185,174,255
MAGNET=MAGNT,198,197,210
MANABOND=MANA,0,255,0
MANA=MANA,0,237,152
MANNEQUIN=MANQN,206,199,162
MATRIX=MATRX,183,174,255
MEMORY=MEMRY,174,183,255
MIDWAY=MDWAY,84,101,207
MINE=MINE,255,100,127
MINING=MINNG,184,201,207
MIRE=MIRE,153,209,199
MUSTER=MUSTR,235,196,0
NECRODERMIS=NECRO,80,209,250
NEST=NEST,80,80,50
NET=NET,0,221,251
OIL=OIL,99,102,106
OMEN=OMEN,255,178,120
ORE=ORE,200,201,163
PAGE=PAGE,218,195,162
PAIN=PAIN,255,108,111
PALLIATION=PALLI,114,243,250
PETAL=PETAL,255,162,216
PHYLACTERY=PHYLA,117,219,153
PHYRESIS=PHYRE,125,97,128
PIN=PIN,194,196,233
PLAGUE=PLGUE,94,226,25
PLAN=PLAN,20,35,235
PLOT=PLOT,255,172,133
POINT=POINT,153,255,130
POLYP=POLYP,236,185,198
POSSESSION=POSSN,60,65,85
PRESSURE=PRESS,255,164,159
PREY=PREY,240,0,0
RALLY=RALLY,25,230,225
REJECTION=REJECT,212,235,242
RELEASE=RELEASE,200,210,50
REPRIEVE=REPR,240,120,50
REVIVAL=REVIVL,130,230,50
REV=REV,255,108,111
RIBBON=RIBBON,233,245,232
RITUAL=RITUAL,155,17,30
ROPE=ROPE,239,223,187
SCREAM=SCREM,0,220,255
SCROLL=SCRLL,206,199,162
SHIELD=SHLD,202,198,186
SHRED=SHRED,255,165,152
SILVER=SILVER,192,192,192
SKEWER=SKEWER,202,192,156
SLIME=SLIME,101,220,163
SOOT=SOOT,211,194,198
SOUL=SOUL,243,190,247
SPITE=SPITE,0,218,255
SPORE=SPORE,122,218,150
STASH=STASH,248,191,0
STORAGE=STORG,255,177,121
STORY=STORY,180,72,195
STRIFE=STRFE,255,89,223
STUDY=STUDY,226,192,165
SUPPLY=SPPLY,70,105,60
TAKEOVER=TKVR,63,49,191
TASK=TASK,191,63,49
THEFT=THEFT,255,176,125
TIDE=TIDE,0,212,187
TIME=TIME,255,121,255
TOWER=tower,0,239,255
TRAP=TRAP,255,121,86
TREASURE=TRSUR,255,184,0
UNITY=UNITY,242,156,255
UNLOCK=UNLCK,222,146,205
VALOR=VALOR,252,250,222
VELOCITY=VELO,255,95,138
VERSE=VERSE,0,237,155
VITALITY=VITAL,255,94,142
VORTEX=VORTX,142,200,255
VOYAGE=VOYAGE,38,150,137
WINCH=WINCH,208,195,203
WIND=WIND,0,236,255
WISH=WISH,255,85,206
WRECK=WRECK,208,55,255

[Neutral]
AWAKENING=AWAKE,0,231,79
GHOSTFORM=GHSTF,223,0,254
MANIFESTATION=MNFST,104,225,8
PETRIFICATION=PETRI,185,201,208
QUEST=QUEST,251,189,0
TRAINING=TRAIN,220,201,0

[Negative]
AGE=AGE,255,137,57
BRIBERY=BRIBE,172,201,235
DOOM=DOOM,255,104,118
FINALITY=FINAL,255,255,255
GOLD=GOLD,248,191,0
INCARNATION=INCRN,247,206,64
MUSIC=MUSIC,255,138,255
PARALYZATION=PRLYZ,220,201,0
PUPA=PUPA,0,223,203
RUST=RUST,255,181,116
SHELL=SHELL,190,207,111
SLEEP=SLEEP,178,192,255
SLEIGHT=SLGHT,185,174,255
SLUMBER=SLMBR,178,205,255
STUN=STUN,226,192,165
WAGE=WAGE,242,190,106
```

**Comment by Hanmac** [comment]
hm i might need to do something with `ACQUIREDTASTE` :/

**Comment by Jetz72** [comment]
Could counter types just have spaces in their names? They don't have to be symbol names anymore, and I think `PutCounter` splits counter lists by comma instead of space...

File: forge-game/src/main/java/forge/game/card/CounterListType.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public static CounterListType get(String s) {
+        return sMap.get(s);
+    }
+
+    public static Collection<CounterListType> getValues() {
+        return sMap.values();
+    }
+
+    public static void add(CounterAiCategory aiCategory, String name, String desc, int red, int green, int blue) {
+        sMap.put(desc, new CounterListType(aiCategory, StringUtils.capitalize(name), desc, red, green, blue));
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/card/CounterListType.java:24]: i noticed i also has a switchup with the params
`name` should be the key for the map

**Comment by jamincollins** [comment]
## Findings
### Logic Correctness and MTG Rule Compliance
#### 1. CounterListType.parseTypes lacks input validation and will crash on malformed CounterLists.txt
The parser splits each line with `line.split("=")` and then `k[1].split(",")` without checking array bounds or numeric format. A malformed line such as a missing `=`, missing colour components, or non-numeric colour values will throw `ArrayIndexOutOfBoundsException` or `NumberFormatException` during `FModel` dynamic data load, crashing the game startup.
Test case:
```java
// Simulate bad line in CounterLists.txt
List<String> bad = List.of("BADLINEWITHOUTEQUALS", "NAME=DESC,notanumber,0,0");
CounterListType.parseTypes(CounterAiCategory.Positive, bad);
// Throws ArrayIndexOutOfBoundsException on first line, NumberFormatException on second
```
Impact: A single typo in the data file prevents Forge from loading. No graceful degradation.

#### 2. CounterAiCategory.valueOf(sectionName) is unsafe
`FModel` loads counters with:
```java
CounterListType.parseTypes(CounterAiCategory.valueOf(sectionName), contents.get(sectionName));
```
If a section header in `CounterLists.txt` is miss-spelled, contains trailing spaces, or uses a different case, `valueOf` throws `IllegalArgumentException` and aborts loading.
Test case:
```java
CounterAiCategory.valueOf("Positive "); // IllegalArgumentException
```
Impact: Data file changes require exact enum name match; no validation or fallback.

### Edge Cases, Null Safety, Boundary Values
#### 3. CounterListType.add stores by `desc` but `getName()` returns capitalised `name`
`add` does `sMap.put(desc, new CounterListType(..., StringUtils.capitalize(name), desc, ...))`. Lookup is by description, yet `getName()` returns the capitalised input name. For entry `ACQUIREDTASTE=ATASTE,139,19,69` the map key is `ATASTE` while `getName()` returns `Acquiredtaste`. Existing code that relied on `CounterEnumType.getName()` returning the enum constant name will now see a different string.
Test case:
```java
CounterListType.parseTypes(CounterAiCategory.Positive, List.of("ACQUIREDTASTE=ATASTE,139,19,69"));
CounterType ct = CounterType.getType("ATASTE");
assert ct.getName().equals("Acquiredtaste"); // true
// Previously enum gave name "acquired taste"
```
Impact: UI labels, logs, and any string comparison on counter names will change silently.

#### 4. No trimming of lines / keys
`parseTypes` does not trim whitespace. A line like ` ICE =ICE,0,239,255` will be parsed with key `" ICE "` and description `"ICE"` leading to a map entry with a leading space key. Lookup via `CounterType.getType("ICE")` will miss it and fall back to enum.
Test case:
```java
List<String> content = List.of(" ICE =ICE,0,239,255");
CounterListType.parseTypes(CounterAiCategory.Positive, content);
assert CounterListType.get("ICE") == null; // true, key is "ICE" with leading space
```
Impact: Silent miss, counters appear missing.

### Data Integrity, Backwards Compatibility
#### 5. Shadowing between list and enum counters
`CounterType.getType` now checks `CounterListType.get(name)` before `CounterEnumType.getType(name)`. If a list entry uses the same description as an existing enum counter's display name, the list entry shadows the enum. Colour/ AI category values diverge.
Test case:
```java
// If CounterLists.txt contains LOYALTY=LOYAL,198,198,198
CounterType ct = CounterType.getType("LOYAL");
 // Returns CounterListType, not CounterEnumType.LOYALTY
```
Impact: Behavioural change without explicit migration; enum fields like `is(CounterEnumType)` will always be false for list instances.

#### 6. CounterEnumType retained a minimal set, breaking assumptions about completeness
Many counters removed from the enum are now only in the list. Code that iterates `CounterEnumType.values()` will no longer see them, e.g. `CounterType.getValues()` now mixes enum and list, but any legacy code enumerating only the enum will miss the moved counters.
Test case:
```java
long count = java.util.Arrays.stream(CounterEnumType.values()).count();
// Before PR ~200, after PR ~15
```
Impact: Backwards compatibility break for modules that expected enum completeness.

### Side Effects & Interactions
#### 7. `countersLoaded` flag prevents re-load but never resets on parse failure
`FModel` sets `countersLoaded = true` after the loop without try/catch. If parsing throws, the flag remains false and the next load attempt will retry, potentially looping. If the exception is swallowed upstream, the flag may stay false forever.
Test case:
```java
// Simulate parse exception
try { CounterListType.parseTypes(...); } catch (Exception e) {}
// countersLoaded remains false, next load repeats
```
Impact: Unclear failure mode, difficult debugging.

### Performance and Security
#### 8. Static mutable map without synchronization
`sMap` is a static `LinkedHashMap` mutated at load time via `add`. In a multi-threaded environment (e.g. parallel mod loading) concurrent `parseTypes` calls can corrupt the map.
Test case:
```java
// Two threads calling parseTypes concurrently
// Possible ConcurrentModificationException or lost entries
```
Impact: Rare but possible data corruption on startup under concurrency.

## Verdict
Request Changes

The PR successfully externalises counters to a file, but the parser lacks validation, error handling, and consistent key semantics. The name/description split, untrimmed input, and unsafe `valueOf` usage make the load fragile and change observable counter names. Backwards compatibility assumptions around `CounterEnumType` completeness and `getName()` values are broken without documentation.

## Recommended Fixes
1. Add robust parsing in `CounterListType.parseTypes`: trim lines, validate split length, catch `NumberFormatException`, log and skip bad entries instead of crashing. Wrap `FModel` loading in try/catch with fallback.
2. Normalize keys: store and lookup by canonical description lower-cased, and make `getName()` return the original name consistently. Consider using `desc` for both lookup and display name to match previous enum behaviour.
3. Validate section names before `valueOf`; provide a clear error message for unknown categories.
4. Add unit tests for malformed lines, duplicate keys, and shadowing scenarios.
5. Document the name/description mapping change and update any UI code that compares counter names.

---
Reviewed by Hermes Agent - muse-glimmer-30b
